### PR TITLE
Avoid folding mutated buffer storages

### LINKED ISSUE #165444: AOTInductor not updating buffers inplace
Hey all, 

I'd like to double check whether updating buffers inplace is currently supported with AOTInductor? Based on the answers on this issue https://github.com/pytorch/pytorch/issues/159124 I think it should be, but it does not seem to work when I load the module from file. If not, is there any workaround we can use at this time (short of making the function pure)? I'm currently on libtorch 2.8.0.

```
class DummyModel(torch.nn.Module):
  def __init__(self):
    super().__init__()
    self.register_buffer("counter", torch.ones(1))
  
  def forward(self):
    self.counter = self.counter + 1.0
    return self.counter
  
ep = torch.export.export(DummyModel(), tuple())
so_path = torch._inductor.aoti_compile_and_package(
    ep,
    inductor_configs={"always_keep_tensor_constants": True}
)
loaded_module = torch._inductor.aoti_load_package(so_path)

print(ep.module()())
print(ep.module()())
print(ep.module()())
print(loaded_module())
print(loaded_module())
print(loaded_module())
```
```
tensor([2.])
tensor([3.])
tensor([4.])
tensor([2.])
tensor([2.])
tensor([2.])
```
@desertfire @ezyang 

cc @chauhang @penguinwu

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185077

AOTInductor packages unlift exported module buffers into runtime-owned constants. For buffer mutation, export re-inserts a copy_ from the functional mutation output back into that buffer. The uniform-value constant folder could fold a singleton get_attr buffer into a compile-time full(...) value before that copy_, so generated packages wrote the same compile-time value into the buffer on every call. Loaded packages therefore returned 2, 2, 2 instead of preserving the counter state as 2, 3, 4.

Track storages written by mutable operator schema arguments and skip uniform-value folding for any node whose fake tensor storage aliases one of those writes. This keeps mutable buffers and their views as runtime reads while preserving uniform folding for immutable values. The alternative would be to special-case AOTI buffer constants, but storage-based alias tracking matches the actual hazard and also covers views of mutated buffers.

The regression test compiles and loads an AOTI package, then calls it repeatedly for assignment, add_, slice assignment, shape-(2,) add_, and indexed mutation with both always_keep_tensor_constants settings.

Fixes #165444
Generated by my agent

Test Plan:
- python test/inductor/test_aot_inductor_package.py TestAOTInductorPackage_cpu.test_buffer_mutations_persist_across_package_calls
- python -m pytest test/inductor/test_aot_inductor_package.py::TestAOTInductorPackage_cpu::test_buffer_mutations_persist_across_package_calls -q
- python -m py_compile torch/_inductor/fx_passes/joint_graph.py test/inductor/test_aot_inductor_package.py
- lintrunner -a torch/_inductor/fx_passes/joint_graph.py test/inductor/test_aot_inductor_package.py

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

File: torch/_inductor/fx_passes/joint_graph.py [hunk 1]
```diff
@@ -326,6 +328,41 @@ def is_zero_int(arg: Any) -> bool:
...
+
+                if idx < len(node.args):
+                    arg = node.args[idx]
+                elif schema_arg.name in node.kwargs:
+                    arg = node.kwargs[schema_arg.name]
+                else:
+                    continue
+
+                pytree.tree_map_only(torch.fx.Node, add_mutated_storage, arg)
+
```
Comment (eellison) [inline torch/_inductor/fx_passes/joint_graph.py]: Can you look around, and see if we have any helpers for this ?

File: torch/_inductor/fx_passes/joint_graph.py [hunk 2]
```diff
@@ -326,6 +328,41 @@ def is_zero_int(arg: Any) -> bool:
...
 
+    def _collect_mutated_storages(self) -> OrderedSet[int]:
+        mutated_storages: OrderedSet[int] = OrderedSet()
+
+        def add_mutated_storage(arg: torch.fx.Node) -> None:
+            storage = get_node_storage(arg)
+            if storage is not None:
+                mutated_storages.add(storage)
+
+        for node in self.module.graph.nodes:  # type: ignore[union-attr]
```
Comment (eellison) [inline torch/_inductor/fx_passes/joint_graph.py]: Thinking aloud.. is there a way to avoid another graph pass here? say like, how does the find_nodes work ? we must already have a mapping of node targets -> nodes. so we could just look through all of them and mark the mutable ones, then, find those nodes.

**Review by eellison — APPROVED** [review]
looks good, one comment.

File: torch/_inductor/fx_passes/joint_graph.py (hunk 1 above)
Comment (jansel) [reply torch/_inductor/fx_passes/joint_graph.py]: My agent says I found and switched this to the existing `torch._library.utils.zip_schema` helper, so the mutable-argument lookup now uses the same schema binding path as the rest of the codebase.

File: torch/_inductor/fx_passes/joint_graph.py (hunk 2 above)
Comment (jansel) [reply torch/_inductor/fx_passes/joint_graph.py]: My agent says I checked `Graph.find_nodes` for this case. It requires a concrete target for `call_function`, so it cannot enumerate all mutable call-function nodes without reaching into private lookup-table internals. I kept the graph scan and applied the helper cleanup above, which removes the bespoke schema argument handling while preserving the mutation tracking.

Comment (eellison) [reply torch/_inductor/fx_passes/joint_graph.py]: disagree, you can just go through the keys of the graph table: https://github.com/pytorch/pytorch/blob/5a5e50f0f8aa36707597e9242652d1a38af09cf1/torch/fx/graph.py#L1345-L1348

Comment (jansel) [reply torch/_inductor/fx_passes/joint_graph.py]: My agent says I updated this to use the FX graph lookup table: mutation collection now walks `graph._find_nodes_lookup_table.table` for call-function targets and calls `find_nodes(..., sort=False)` only for mutable `OpOverload` targets. That removes the full graph node scan while keeping the storage-alias write tracking intact.

**Comment by jansel** [comment]
My agent says the current Dr.CI inductor benchmark failures appear preexisting rather than caused by this PR. The reported models fail in CUDA training with `eager_two_runs_differ` before Dynamo/Inductor compilation (`Compilation time ... 0`, `Dynamo produced 0 graphs`), and the same failures reproduce on merge-base workflow run https://github.com/pytorch/pytorch/actions/runs/27981186756 for `tf_efficientnet_b0` and `shufflenet_v2_x1_0`/related TorchBench models. I left the PR code unchanged.
